src/Core.hs

```diff
@@ -35,14 +37,20 @@ newtype MacroVar = MacroVar Unique
 instance Show MacroVar where
   show (MacroVar i) = "(MacroVar " ++ show (hashUnique i) ++ ")"
 
-data Pattern
-  = PatternIdentifier Ident Var
-  | PatternEmpty
-  | PatternCons Ident Var Ident Var
-  | PatternList [(Ident, Var)]
-  | PatternAny
+data ConstructorPattern
+  = ConstructorPattern !Constructor [(Ident, Var)]
+  | AnyConstructor Ident Var
```

Comment:
Checking my understanding: Is this the `a` in something like
```haskell
   case x of
     a -> a -- you've learned nothing about x
```

Reply:
Yes! It generalizes catchall patterns.